<?xml version="1.0"?>
<launch>
  ###########  launch config  ###########
  <arg name="real_machine" default="True" />
  <arg name="simulation" default="False" />
  <arg name="control_mode" default="0" />
  <arg name="estimate_mode"  default= "1" />
  <arg name="sim_estimate_mode"  default= "2" />
  <arg name="headless" default="True" />
  <arg name="worldtype" default="$(find aerial_robot_simulation)/gazebo_model/world/empty.world" />
  <arg name="launch_gazebo" default="True" />
  <arg name="direct_model" default="False" />
  <arg name="direct_model_name" />
  <arg name="spawn_x" default="0.0"/>
  <arg name="spawn_y" default="0.0"/>
  <arg name="spawn_z" default="0.0"/>
  <arg name="spawn_yaw" default="0.0"/>
  <arg name="config_dir" default="$(find beetle)/config"/>
  <arg name="robot_id" default="1" />
  <arg name="robot_ns" value="beetle$(arg robot_id)" />

  ###########  Parameters  ###########
  <group ns="$(arg robot_ns)">
    # EGOMOTION_ESTIMATE = 0
    # EXPERIMENT_ESTIMATE = 1. for unstable mocap, use this mode
    # GROUND_TRUTH = 2

    ###########  Basic Param  ###########
    <param name="estimation/mode" value= "$(arg sim_estimate_mode)" if="$(arg simulation)" />
    <param name="estimation/mode" value= "$(arg estimate_mode)" if="$(arg real_machine)"/>
    <param name = "robot_id" value="$(arg robot_id)"/>
   ###########  Robot Model  ###########
   <rosparam file="$(arg config_dir)/RobotModel.yaml" command="load" />

   ###########  Motor Config  ###########
   <rosparam file="$(arg config_dir)/MotorInfoDShot.yaml" command="load" />

   ###########  Servo Config  ###########
   <rosparam file="$(arg config_dir)/Servo.yaml" command="load"/>

   ###########  Battery Config  ###########
   <rosparam file="$(arg config_dir)/Battery.yaml" command="load" />

   ###########  Control  ###########
   <rosparam file="$(arg config_dir)/BeetleNMPCFull.yaml" command="load" />

   ###########  Sensor Fusion  ###########
   <rosparam file="$(arg config_dir)/StateEstimation.yaml" command="load" />
   <rosparam file="$(arg config_dir)/Simulation.yaml" command="load" if="$(eval arg('simulation') and not arg('real_machine'))"/>

   ###########  Navigation  ###########
   <param name="flight_navigation_plugin_name" value="aerial_robot_navigation/beetle_navigation" />
   <rosparam file="$(arg config_dir)/NavigationConfig.yaml" command="load" />

  </group>

  ###########  Base Platform  ###########
  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node" ns="$(arg robot_ns)" output="screen" >
    <param name="tf_prefix" value="$(arg robot_ns)"/>
    <param name="param_verbose" value="false"/>
    <param name="main_rate" value="100"/>
  </node>

  ###########  Robot Model  ###########
  <arg name="description_mode" value="urdf" unless="$(arg simulation)" />
  <arg name="description_mode" value="gazebo" if="$(arg simulation)" />
  <include file="$(find aerial_robot_model)/launch/aerial_robot_model.launch" >
    <arg name="headless" value="$(arg headless)" />
    <arg name="robot_model" value="$(find beetle)/robots/beetle-art.$(arg description_mode).xacro"/>
    <arg name="robot_ns" value="$(arg robot_ns)" />
    <arg name="rviz_config" value="$(arg config_dir)/nmpc.rviz" />
    <arg name="rviz_init_pose" value="$(arg config_dir)/RvizInit.yaml" />
    <arg name="need_joint_state" value="false" if ="$(eval arg('simulation') or arg('real_machine'))"/>
  </include >

  ##########  Sensors  ##########
  <include file="$(find beetle)/launch/include/sensors.launch.xml" >
    <arg name="real_machine" value="$(arg real_machine)" />
    <arg name="simulation" value="$(arg simulation)" />
    <arg name="robot_ns" value="$(arg robot_ns)" />
    <arg name="robot_id" value="$(arg robot_id)"/>
  </include >

  ###########  Servo Bridge  ###########
  <node pkg="aerial_robot_model" type="servo_bridge_node" name="servo_bridge" ns="$(arg robot_ns)" output="screen" />

  ########## Simulation in Gazebo #########
  <include file="$(find aerial_robot_simulation)/launch/simulation.launch" if = "$(eval arg('simulation') and not arg('real_machine'))" >
    <arg name="robot_ns" default="$(arg robot_ns)" />
    <arg name="gui" value="false" if="$(arg headless)"/>
    <arg name="headless" default="$(arg headless)" />
    <arg name="launch_gazebo" default="$(arg launch_gazebo)" />
    <arg name="worldtype" value="$(arg worldtype)" />
    <arg name="spawn_x" value="$(arg spawn_x)" />
    <arg name="spawn_y" value="$(arg spawn_y)" />
    <arg name="spawn_z" value="$(arg spawn_z)" />
    <arg name="spawn_yaw" value="$(arg spawn_yaw)" />
  </include>

</launch>
